### DISCUSSION TITLE
Toolbar menu problem when exporting and loading project data(JSON)

### DISCUSSION THREAD

**Discussion by alifsuria** [GrapesJS/grapesjs#5047] [Q&A]
Hello all,

I need help regarding the toolbar menu

This is the default grapeJS toolbar menu
<img width="488" alt="Default" src="https://user-images.githubusercontent.com/35814118/231660911-bfa36e56-7398-4431-a3da-a8ede179c03b.png">

This is a toolbar menu that was customized in a component.model
<img width="277" alt="Custom" src="https://user-images.githubusercontent.com/35814118/231660936-a08d5233-f271-46d3-b425-a83d26e87d34.png">

When I export the project data using editor.getProjectData() and load it, The custom toolbar menu is not added and is replaced by the default one.

Please help

**Comment by artf** [comment]
Are you properly defining the custom component in a plugin? Can you show the related code?

**Comment by alifsuria** [comment]
Thank you @artf for replying.

This is how I define a new custom component in a  plugin.

    `editor.Components.addType('package-link', {
        model: {
            defaults: {
                tagName: 'section',
                components: {
                    name: 'accordion-list-box',
                    tagName: 'div',
                    components: {
                        content: '<p>Hello World</p>',
                        toolbar: [{ command: 'moveItem', label: `<span><i /class="fa-solid fa-arrows"></i></span>` }, { command: 'cloneItem', label: `<span><i class="fa-solid fa-copy"></i></span>` }, { command: 'deleteItem', label: `<span><i class="fa-solid fa-trash"></i></span>` },],
                    },
                    toolbar: [{ command: 'moveItem', label: `<span><i /class="fa-solid fa-arrows"></i></span>` }, { command: 'cloneItem', label: `<span><i class="fa-solid fa-copy"></i></span>` }, { command: 'deleteItem', label: `<span><i class="fa-solid fa-trash"></i></span>` },],
                },
            },
        },
    })
    editor.Blocks.add("Package Link", {
        label: "Package Link",
        media: `<span class='block-icon'><i class='fa-solid fa-box-open'></i></span>`,
        category: 'Ecommerce',
        content: { type: 'package-link' }
    })`
    
    This is my export code
    `$axios.get(endpoint).then((response) => {
                // let projectData = editor.getProjectData();
                  editor.store().then(() => {
                    if (response.data.length === 0) {
                      $axios.post(endpoint, {data: editor.getProjectData()}).then((response) => {
                        console.log(response);
                      });
                    } else {
                      let getExistingData = response.data[0];
                      console.log(getExistingData);
                      axios.put(
                        `${endpoint}/${getExistingData.id}`,
                        {data:editor.getProjectData()}
                      ).then((data)=>{
                        console.log(data)
                        console.log('data updated')
                      });
                    }
                  });
                }).catch((err)=>{
                  console.log(err)
                });`
I'm sorry for taking your time about this issue.

**Comment by artf** [comment]
As toolbar property is not stored directly in the project you should create a custom component containing your toolbar configuration so when it's initialised it will take the right configuration.
You can also [extend](https://grapesjs.com/docs/modules/Components.html#extend-component-type) default components with your new toolbar if this is something changing on more than one custom component.

**Reply by alifsuria** [reply]
The toolbar only stayed when adding it on the most 'top-level/parent' but if we added it deep into the child or specific component or element. It doesn't stay. Does it was undoable?

`editor.Components.addType('test-box', {
        model: {
            defaults: {
                tagName: 'div',
                // This toolbar will persist even after saving and loading the project data
                toolbar: [{ command: 'moveItem', label: `<span><i /class="fa-solid fa-arrows"></i></span>` }, { command: 'cloneItem', label: `<span><i class="fa-solid fa-copy"></i></span>` }, { command: 'deleteItem', label: `<span><i class="fa-solid fa-trash"></i></span>` },],
                components: [
                    {
                        tagName: "div",
                        components: {
                            tagName: 'div',
                            name: 'outer-box',
                            components: [{
                                tagName: 'div',
                                name: 'first-column'
                            }, {
                                // This toolbar didn't persist even after saving and loading the project data it only shows when added to the editor
                                toolbar: [{ command: 'moveItem', label: `<span><i /class="fa-solid fa-arrows"></i></span>` }, { command: 'cloneItem', label: `<span><i class="fa-solid fa-copy"></i></span>` }, { command: 'deleteItem', label: `<span><i class="fa-solid fa-trash"></i></span>` },],
                                tagName: 'div',
                                name: 'second-column',
                            }]
                        }
                    },
                    {
                        tagName: "p",
                        components: `<p>Hello World</p>`
                    }
                ]
            }
        }
    });`
